### PR TITLE
feat(agent, CodeAct 2.2): native CodeAct support for Browsing

### PR DESCRIPTION
**Pull request by xingyaoww**
**End-user friendly description of the problem this fixes or functionality that this introduces**

- [X] Include this change in the Release Notes. If checked, you must provide an **end-user friendly** description for your change below

Allow CodeAct agent to perform web browsing by itself (no more delegation to BrowsingAgent).

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

- Attach BrowserGym's action space (accessibility tree) to CodeAct directly
- Add integration test to make sure it is working

Example traj:
```
Look at https://github.com/All-Hands-AI/OpenHands/pull/8, and tell me what is happening there and what did @asadm suggest.

**BrowseInteractiveAction**
THOUGHT: I'll help you look at that GitHub pull request. Let me navigate to the URL.
BROWSER_ACTIONS: goto('https://github.com/All-Hands-AI/OpenHands/pull/8')

15:29:38 - openhands:DEBUG: stream.py:148 - Adding BrowserOutputObservation id=4 from AGENT
15:29:38 - openhands:DEBUG: agent_controller.py:135
OBSERVATION
[Agent Controller default] **BrowserOutputObservation**
URL: https://github.com/All-Hands-AI/OpenHands/pull/8
Error: False
Open pages: ['https://github.com/All-Hands-AI/OpenHands/pull/8']
Active page index: 0
Last browser action: goto('https://github.com/All-Hands-AI/OpenHands/pull/8')
Last browser action error: 
Focused element bid: 149
AX tree: RootWebArea 'Add MIT LICENSE by asadm · Pull Request #8 · All-Hands-AI/OpenHands · GitHub', focused
        [152] link 'Skip to content', clickable
        [170] banner ''
                [171] heading 'Navigation Menu'
                [183] link 'Homepage', clickable
                [189] navigation 'Global'
                        [190] list ''
                                [191] listitem ''
                                        [192] button 'Product', expanded=False
                                [253] listitem ''
                                        [254] button 'Solutions', expanded=False
                                [295] listitem ''
                                        [296] button 'Resources', expanded=False
                                [328] listitem ''
                                        [329] button 'Open Source', expanded=False
                                [354] listitem ''
                                        [355] button 'Enterprise', expanded=False
                                [384] listitem ''
                                        [385] link 'Pricing', clickable
                [388] button 'Search or jump to…', clickable, hasPopup='dialog'
                [510] link 'Sign in', clickable
                [511] link 'Sign up', clickable
        [530] main ''
                [537] link 'All-Hands-AI', clickable
                StaticText '/'
                [539] strong ''
                        [540] link 'OpenHands', clickable
                StaticText 'Public'
                [544] list ''
                        [545] listitem ''
                                [546] link 'You must be signed in to change notification settings', clickable, describedby='tooltip-56ff1a9d-4cda-4907-860a-9064338030c4'
                                        StaticText 'Notifications'
                        [549] listitem ''
                                [550] link 'Fork 3.8k', clickable
                                        StaticText '3.8k'
                        [553] listitem ''
                                [555] link 'You must be signed in to star a repository', clickable
                                        StaticText 'Star'
                                        StaticText '33.3k'
                [560] navigation 'Repository'
                        [561] list ''
                                [562] listitem ''
                                        [563] link 'Code', clickable
                                                StaticText 'Code'
                                [567] listitem ''
                                        [568] link 'Issues 141', clickable
                                                StaticText 'Issues'
                                [572] listitem ''
                                        [573] link 'Pull requests 35', clickable
                                                StaticText 'Pull requests'
                                [577] listitem ''
                                        [578] link 'Discussions', clickable
                                                StaticText 'Discussions'
                                [582] listitem ''
                                        [583] link 'Actions', clickable
                                                StaticText 'Actions'
                                [587] listitem ''
                                        [588] link 'Projects 2', clickable
                                                StaticText 'Projects'
                                [592] listitem ''
                                        [593] link 'Security', clickable
                                                StaticText 'Security'
                                [596] listitem ''
                                        [597] link 'Insights', clickable
                                                StaticText 'Insights'
                [656] Details ''
                        [657] button 'New issue', clickable, expanded=False
                [675] heading 'Add MIT LICENSE #8'
                        StaticText 'Add MIT LICENSE'
                        StaticText '#8'
                [683] link 'huybery', clickable
                StaticText 'merged 1 commit into'
                [685] link 'All-Hands-AI:main', clickable
                        StaticText 'All-Hands-AI'
                        StaticText 'main'
                StaticText 'from'
                [690] link 'asadm:patch-1', clickable
                        StaticText 'asadm'
                        StaticText 'patch-1'
                button 'Copy'
                [697] generic, live='polite', atomic, relevant='additions text'
                StaticText 'on Mar 16'
                StaticText '+25'
                StaticText '−0'
                [738] navigation 'Pull request tabs'
                        [739] link 'Conversation 5', clickable
                        [742] link 'Commits 1', clickable
                        [745] link 'Checks 0', clickable
                        [748] link 'Files changed 1', clickable
                [754] heading 'Conversation'
                [761] link 'asadm', clickable
                        [762] image 'asadm'
                [767] Details '', clickable
                        [768] button 'Show options', clickable, hasPopup='menu', expanded=False
                                [771] image 'Show options'
                StaticText 'Contributor'
                [777] heading 'asadm commented on Mar 14'
                        [780] strong ''
                                [781] link 'asadm', clickable
                                StaticText ''
                        StaticText 'commented'
                        [782] link 'on Mar 14', clickable
                                StaticText 'on Mar 14'
                [786] paragraph ''
                        StaticText 'Before we began, we should decide on a license. If this is "open", I would consider MIT or Apache 2.0 to be safe choices.'
                [787] Section ''
                [798] Section ''
                        [803] button 'thumbs up (2): asadm, 07:11AM on March 14', disabled=True, pressed='false', describedby='tooltip-bc724ba6-b267-4393-869d-1b1932b3f802'
                                [804] image '+1'
                                StaticText '2'
                        [807] button 'thumbs down (4): asadm, 07:11AM on March 14', disabled=True, pressed='false', describedby='tooltip-4da75324-5eee-4d45-ab2d-1c5ca2725b86'
                                [808] image '-1'
                                StaticText '4'
                [836] link '@asadm', clickable
                        [837] image '@asadm'
                [839] code ''
                        [840] link 'Create MIT LICENSE', clickable
                [842] button 'Verified'
                        StaticText 'Verified'
                [867] code ''
                        [868] link '70b4ebf', clickable
                [872] link '@jjfantini', clickable
                        [873] image '@jjfantini'
                [878] Details '', clickable
                        [879] button 'Show options', clickable, hasPopup='menu', expanded=False
                                [882] image 'Show options'
                [886] heading 'jjfantini commented on Mar 14'
                        [888] strong ''
                                [889] link 'jjfantini', clickable
                                StaticText ''
                        StaticText 'commented'
                        [890] link 'on Mar 14', clickable
                                StaticText 'on Mar 14'
                [892] LayoutTable ''
                        [894] LayoutTableRow ''
                                [895] LayoutTableCell 'what about an open license that restricts commercial use? to keep it for the people?'
                                        [896] paragraph ''
                                                StaticText 'what about an open license that restricts commercial use? to keep it for the people?'
                [900] Section ''
                        [906] button 'react with thumbs up', disabled=True, pressed='false', describedby='tooltip-b65a16ac-f9df-4931-a728-0b5b0f98ac3b'
                                [907] image '+1'
                                StaticText '1'
                [919] Section ''
                [931] link '@pombredanne', clickable
                        [932] image '@pombredanne'
                [937] Details '', clickable
                        [938] button 'Show options', clickable, hasPopup='menu', expanded=False
                                [941] image 'Show options'
                [945] heading 'pombredanne commented on Mar 15'
                        [947] strong ''
                                [948] link 'pombredanne', clickable
                                StaticText ''
                        StaticText 'commented'
                        [949] link 'on Mar 15', clickable
                                StaticText 'on Mar 15'
                [951] LayoutTable ''
                        [953] LayoutTableRow ''
                                [954] LayoutTableCell 'Then this is proprietary, not open source.'
                                        [955] blockquote ''
                                                [956] paragraph ''
                                                        StaticText 'what about an open license that restricts commercial use? to keep it for the people?'
                                        [957] paragraph ''
                                                StaticText 'Then this is proprietary, not open source.'
                [961] Section ''
                        [967] button 'react with thumbs up', disabled=True, pressed='false', describedby='tooltip-85b71239-9ad3-4abf-a6b0-3a75a3699ea5'
                                [968] image '+1'
                                StaticText '1'
                [980] Section ''
                [992] link '@asadm', clickable
                        [993] image '@asadm'
                [998] Details '', clickable
                        [999] button 'Show options', clickable, hasPopup='menu', expanded=False
                                [1002] image 'Show options'
                StaticText 'Contributor'
                StaticText 'Author'
                [1010] heading 'asadm commented on Mar 15'
                        [1012] strong ''
                                [1013] link 'asadm', clickable
                                StaticText ''
                        StaticText 'commented'
                        [1014] link 'on Mar 15', clickable
                                StaticText 'on Mar 15'
                [1016] LayoutTable ''
                        [1018] LayoutTableRow ''
                                [1019] LayoutTableCell "If we go with a non-commercial license: many will not contribute at all, since it's not open source in true sense commercial folks will just reimplement ideas here, LoC vs. impact is pretty low in genAI world, will take a weekend to replicate ideas this repo will be left behind."
                                        [1020] paragraph ''
                                                StaticText 'If we go with a non-commercial license:'
                                        [1021] list ''
                                                [1022] listitem ''
                                                        ListMarker '•'
                                                        StaticText "many will not contribute at all, since it's not open source in true sense"
                                                [1023] listitem ''
                                                        ListMarker '•'
                                                        StaticText 'commercial folks will just reimplement ideas here, LoC vs. impact is pretty low in genAI world, will take a weekend to replicate ideas'
                                                [1024] listitem ''
                                                        ListMarker '•'
                                                        StaticText 'this repo will be left behind.'
                [1028] Section ''
                        [1034] button 'react with thumbs up', disabled=True, pressed='false', describedby='tooltip-9bcae276-b28b-4a34-aa56-39a18ca3bf9d'
                                [1035] image '+1'
                                StaticText '3'
                [1047] Section ''
                [1059] link '@hannibalhuang', clickable
                        [1060] image '@hannibalhuang'
                [1065] Details '', clickable
                        [1066] button 'Show options', clickable, hasPopup='menu', expanded=False
                                [1069] image 'Show options'
                [1073] heading 'hannibalhuang commented on Mar 16'
                        [1075] strong ''
                                [1076] link 'hannibalhuang', clickable
                                StaticText ''
                        StaticText 'commented'
                        [1077] link 'on Mar 16', clickable
                                StaticText 'on Mar 16'
                [1079] LayoutTable ''
                        [1081] LayoutTableRow ''
                                [1082] LayoutTableCell 'agree with @asadm here and also just commented on issue #10 , MIT or Apache 2.0 are safer choices than CC-bys'
                                        [1083] paragraph ''
                                                StaticText 'agree with'
                                                [1084] link '@asadm', clickable
                                                StaticText 'here and also just commented on issue'
                                                [1085] link '#10', clickable
                                                StaticText ', MIT or Apache 2.0 are safer choices than CC-bys'
                [1101] Section ''
                [1116] link '@huybery', clickable
                        [1117] image '@huybery'
                [1118] link 'huybery', clickable
                StaticText 'mentioned this pull request'
                [1119] link 'on Mar 16', clickable
                        StaticText 'on Mar 16'
                [1122] link 'Which open source licence? #10', clickable
                        StaticText '#10'
                [1130] link '@huybery', clickable
                        [1131] image '@huybery'
                [1136] Details '', clickable
                        [1137] button 'Show options', clickable, hasPopup='menu', expanded=False
                                [1140] image 'Show options'
                StaticText 'Collaborator'
                [1146] heading 'huybery commented on Mar 16'
                        [1148] strong ''
                                [1149] link 'huybery', clickable
                                StaticText ''
                        StaticText 'commented'
                        [1150] link 'on Mar 16', clickable
                                StaticText 'on Mar 16'
                [1152] LayoutTable ''
                        [1154] LayoutTableRow ''
                                [1155] LayoutTableCell 'Thank you @asadm , more people think the MIT license is a better fit for OpenDevin.'
                                        [1156] paragraph ''
                                                StaticText 'Thank you'
                                                [1157] link '@asadm', clickable
                                                StaticText ', more people think the MIT license is a better fit for OpenDevin.'
                [1161] Section ''
                        [1167] button 'react with heart', disabled=True, pressed='false', describedby='tooltip-ef355302-058e-4025-a6f6-93da25b44ff1'
                                [1168] image 'heart'
                                StaticText '3'
                [1180] Section ''
                [1194] link '@huybery', clickable
                        [1195] image '@huybery'
                [1196] link 'huybery', clickable
                StaticText 'merged commit'
                [1197] link '39add27', clickable
                        [1198] code ''
                                StaticText '39add27'
                StaticText 'into'
                StaticText 'All-Hands-AI'
                StaticText ':'
                StaticText 'main'
                [1203] link 'on Mar 16', clickable
                        StaticText 'on Mar 16'
                [1210] link '@neubig', clickable
                        [1211] image '@neubig'
                [1212] link 'neubig', clickable
                StaticText 'mentioned this pull request'
                [1213] link 'on Mar 16', clickable
                        StaticText 'on Mar 16'
                [1216] link 'LICENSE - Attribution-NonCommercial 4.0 International #12', clickable
                        StaticText '#12'
                [1225] link 'xcodebuild', clickable
                StaticText 'pushed a commit to xcodebuild/OpenDevin that referenced this pull request'
                [1226] link 'on Mar 31', clickable
                        StaticText 'on Mar 31'
                [1233] link '@asadm', clickable
                        [1234] image '@asadm'
                [1236] code ''
                        [1237] link 'Create MIT LICENSE (', clickable
                        [1238] link 'All-Hands-AI#8', clickable
                        [1239] link ')', clickable
                [1243] code ''
                        [1244] link 'a4a2c3f', clickable
                [1250] link '@enyst', clickable
                        [1251] image '@enyst'
                [1252] link 'enyst', clickable
                StaticText 'mentioned this pull request'
                [1253] link 'on Sep 19', clickable
                        StaticText 'on Sep 19'
                [1256] link "[Bug]: LLM retries don't obey Ctrl+C either #3965", clickable
                        StaticText '#3965'
                StaticText '2 tasks'
                [1274] link 'Sign up for free', clickable
                [1275] strong ''
                        StaticText 'to join this conversation on GitHub'
                StaticText '. Already have an account?'
                [1276] link 'Sign in to comment', clickable
                [1280] form 'Select reviewers'
                        StaticText 'Reviewers'
                        StaticText 'No reviews'
                [1286] form 'Select assignees'
                        StaticText 'Assignees'
                        StaticText 'No one assigned'
                StaticText 'Labels'
                StaticText 'None yet'
                [1295] form 'Select projects'
                        StaticText 'Projects'
                        [1299] generic, live='polite', relevant='additions text'
                        StaticText 'None yet'
                [1302] form 'Select milestones'
                        StaticText 'Milestone'
                        StaticText 'No milestone'
                [1308] form 'Link issues'
                        StaticText 'Development'
                        [1312] paragraph ''
                                StaticText 'Successfully merging this pull request may close these issues.'
                        [1314] paragraph ''
                                StaticText 'None yet'
                StaticText '5 participants'
                [1319] link '@asadm', clickable
                        [1320] image '@asadm'
                [1321] link '@jjfantini', clickable
                        [1322] image '@jjfantini'
                [1323] link '@pombredanne', clickable
                        [1324] image '@pombredanne'
                [1325] link '@hannibalhuang', clickable
                        [1326] image '@hannibalhuang'
                [1327] link '@huybery', clickable
                        [1328] image '@huybery'
        [1346] contentinfo ''
                [1347] heading 'Footer'
                [1350] link 'Homepage', clickable
                StaticText '© 2024 GitHub,\xa0Inc.'
                [1353] navigation 'Footer'
                        [1354] heading 'Footer navigation'
                        [1355] list 'Footer navigation'
                                [1356] listitem ''
                                        [1357] link 'Terms', clickable
                                [1358] listitem ''
                                        [1359] link 'Privacy', clickable
                                [1360] listitem ''
                                        [1361] link 'Security', clickable
                                [1362] listitem ''
                                        [1363] link 'Status', clickable
                                [1364] listitem ''
                                        [1365] link 'Docs', clickable
                                [1366] listitem ''
                                        [1367] link 'Contact', clickable
                                [1368] listitem ''
                                        [1369] button 'Manage cookies', clickable
                                [1370] listitem ''
                                        [1371] button 'Do not share my personal information', clickable
        [1381] generic, live='polite', atomic, relevant='additions text'
        [1382] generic, live='assertive', atomic, relevant='additions text'



==============
[Agent Controller default] LEVEL 0 LOCAL STEP 1 GLOBAL STEP 1

15:29:51 - openhands:DEBUG: llm.py:405 - Cost: 0.03 USD | Accumulated Cost: 0.08 USD
Input tokens: 7993 | Output tokens: 384

15:29:51 - openhands:DEBUG: agent_controller.py:135
ACTION
[Agent Controller default] **MessageAction** (source=None)
CONTENT: Based on the GitHub pull request, here's what happened:

1. @asadm created a pull request (#8) to add an MIT LICENSE to the OpenHands repository.

2. In his initial comment, @asadm suggested:
   > "Before we began, we should decide on a license. If this is 'open', I would consider MIT or Apache 2.0 to be safe choices."

3. There was a discussion about licensing where:
   - @jjfantini suggested considering a license that restricts commercial use
   - @pombredanne pointed out that restricting commercial use would make it proprietary, not open source
   
4. @asadm made a strong case against non-commercial licenses, explaining that:
   - Many people won't contribute if it's not truly open source
   - Commercial entities would just reimplement the ideas anyway (especially in the GenAI world where code replication is relatively quick)
   - The repository would likely be left behind

5. @hannibalhuang and @huybery agreed with @asadm's suggestion, and ultimately the pull request was merged, adding the MIT license to the repository.

The key insight from @asadm was that choosing a permissive license like MIT or Apache 2.0 would be better for the project's growth and adoption, rather than trying to restrict commercial use. His argument was particularly relevant for AI projects where the barrier to reimplementation is relatively low, and having a truly open license would encourage more community participation and adoption.

The pull request was successful and the MIT license was added to the repository, making it fully open source with very permissive terms for both commercial and non-commercial use.
```

---
**Link of any specific issues this addresses**

---

To run this PR locally, use the following command:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   -e SANDBOX_RUNTIME_CONTAINER_IMAGE=ghcr.io/all-hands-ai/runtime:1c050f9-nikolaik   --name openhands-app-1c050f9   ghcr.io/all-hands-ai/runtime:1c050f9
```

### CODE REVIEW

**Review by tofarr — APPROVED** [review]
LGTM. Can't wait to try it

**Comment by xingyaoww** [comment]
Temporarily convert this to draft as i'm still running some eval to make sure this doesn't degrade our SWE-Bench perf & we are getting some reasonable browsing performance

**Comment by xingyaoww** [comment]
SWE-Bench Lite: 130/299 (43%) on this PR is comparable (and better) to our https://www.swebench.com/ submission (41.67%)
```
## Summary
- submitted instances: 299
- empty patch instances: 3
- resolved instances: 130
- unresolved instances: 169
- error instances: 0
```

Running Miniwob now

**Comment by xingyaoww** [comment]
Miniwob performance (86/125) = 68.8% - this is even better than our Browsing agent in the paper after some bug fixes :)

image

**Comment by enyst** [comment]
> SWE-Bench Lite: 130/299 (43%) on this PR is comparable (and better) to our https://www.swebench.com/ submission (41.67%)

Awesome! 🥳

Why do you think that may be? It's very interesting to see, because if we're adding here a new action, and it isn't used for these tasks, just known to the LLM... it seems counterintuitive that the LLM is doing _better_? Could it be that these 2% are part of the normal variance with tools use, or what could explain it?

File: evaluation/miniwob/run_infer.py [hunk 1]
```diff
@@ -85,8 +104,13 @@ def initialize_runtime(
     logger.info(obs, extra={'msg_type': 'OBSERVATION'})
     goal = obs.content
 
+    action = BrowseInteractiveAction(browser_actions='noop(1000)')
```
Comment (neubig) [inline evaluation/miniwob/run_infer.py:107]: Could you add a comment here about what this is doing?

File: evaluation/miniwob/run_infer.py [hunk 2]
```diff
@@ -47,25 +55,36 @@ def get_config(
...
         workspace_mount_path=None,
     )
+    if metadata.llm_config.log_completions:
+        metadata.llm_config.log_completions_folder = os.path.join(
+            metadata.eval_output_dir, 'llm_completions', env_id
+        )
+        logger.info(
+            f'Logging LLM completions for instance {env_id} to '
+            f'{metadata.llm_config.log_completions_folder}'
+        )
```
Comment (neubig) [inline evaluation/miniwob/run_infer.py:80]: Is there a way we can avoid adding this copypasta to each eval folder and put it in the evaluation utils.py instead?

File: evaluation/swe_bench/run_infer.py [hunk 3]
```diff
@@ -88,6 +89,13 @@ def get_instruction(instance: pd.Series, metadata: EvalMetadata):
             '5. Think about edgecases and make sure your fix handles them as well\n'
             "Your thinking should be thorough and so it's fine if it's very long.\n"
         )
+
+    if RUN_WITH_BROWSING:
+        instruction += (
+            '<IMPORTANT!>\n'
+            'You SHOULD NEVER attempt to browse the web. '
+            '</IMPORTANT!>\n'
+        )
```
Comment (neubig) [inline evaluation/swe_bench/run_infer.py:98]: Is there a way we can keep track of statistics of how often this happened? We're not guaranteed that the agent will follow these directions, so at least being able to post-hoc audit whether it happened or not would be good.

**Review by neubig — APPROVED** [review]
Looks great to me! Just a few small comments that would ideally be resolved before merging (but I don't want to be a blocker).

File: evaluation/miniwob/run_infer.py (hunk 1 above)
Comment (xingyaoww) [reply evaluation/miniwob/run_infer.py:105]: done!

File: evaluation/swe_bench/run_infer.py (hunk 3 above)
Comment (xingyaoww) [reply evaluation/swe_bench/run_infer.py:99]: Yeah! It is not hard - and simple cat with jq should be sufficient

`cat output.jsonl | jq '.history[].action' | sort | uniq -c`

image
Did a quick check - there's no `browse_interactive` action in our SWE-Bench run

**Comment by xingyaoww** [comment]
> Could it be that these 2% are part of the normal variance with tools use, or what could explain it?

@enyst my intuition is that it is likely randomness maybe plus a lot of smaller fixes we've merged in `main` :)

File: evaluation/miniwob/run_infer.py (hunk 2 above)
Comment (xingyaoww) [reply evaluation/miniwob/run_infer.py]: done!

**Comment by jeezrick** [comment]
> Miniwob

This result is really surprising! It's counterintuitive that if you just inject the same browser agent prompt into a larger whole prompt (i.e., the code act agent 2.2 prompt), the browser-related benchmark could be significantly better. So, is the browser part of the prompt the same? And are you testing this with the same LLM model?

**Pull request merged**